src/blocking.rs

```diff
@@ -0,0 +1,646 @@
+    /// individually on each user/group. If rotation is only needed for a specific group, it is strongly recommended
+    /// to call [user_rotate_private_key()](user\/trait.UserOps.html#tymethod.user_rotate_private_key) or
+    /// [group_rotate_private_key()](group\/trait.GroupOps.html#tymethod.group_rotate_private_key) instead.
+    /// # Arguments
+    /// - `rotations` - PrivateKeyRotationCheckResult that holds all users and groups to be rotated
+    /// - `password` - Password to unlock the current user's user master key
+    pub fn rotate_all(
+        &self,
+        rotations: &PrivateKeyRotationCheckResult,
+        password: &str,
+    ) -> Result<(
+        Option<UserUpdatePrivateKeyResult>,
+        Option<Vec<GroupUpdatePrivateKeyResult>>,
+    )> {
+        self.runtime
+            .enter(|| block_on(self.ironoxide.rotate_all(rotations, password)))
+    }
+}
+
+/// Creates a tokio runtime with the default number of core threads (num of cores on a machine)
+/// and an elevated number of blocking_threads as we expect heavy concurrency to be network-bound
+fn create_runtime() -> tokio::runtime::Runtime {
+    tokio::runtime::Builder::new()
+        .threaded_scheduler() // use multi-threaded scheduler
+        .enable_all() // enable both I/O and time drivers
+        .max_threads(250) // core_threads default to number of cores, blocking threads are max - core
+        .build()
+        .expect("tokio runtime failed to initialize")
+}
+
+/// Initialize the IronOxide SDK with a device. Verifies that the provided user/segment exists and the provided device
+/// keys are valid and exist for the provided account. If successful, returns an instance of the IronOxide SDK
+pub fn initialize(device_context: &DeviceContext) -> Result<BlockingIronOxide> {
+    let rt = create_runtime();
+    let maybe_io = rt.enter(|| block_on(crate::initialize(device_context)));
+    maybe_io.map(|io| BlockingIronOxide {
+        ironoxide: io,
+        runtime: rt,
+    })
+}
+
+/// Initialize the IronOxide SDK and check to see if the user that owns this `DeviceContext` is
+/// marked for private key rotation, or if any of the groups that the user is an admin of is marked
+/// for private key rotation.
+pub fn initialize_check_rotation(
+    device_context: &DeviceContext,
+) -> Result<InitAndRotationCheck<BlockingIronOxide>> {
+    let rt = create_runtime();
+    let maybe_init = rt.enter(|| block_on(crate::initialize_check_rotation(device_context)));
+    maybe_init.map(|init| match init {
+        NoRotationNeeded(io) => NoRotationNeeded(BlockingIronOxide {
+            ironoxide: io,
+            runtime: rt,
+        }),
+        RotationNeeded(io, rot) => RotationNeeded(
+            BlockingIronOxide {
+                ironoxide: io,
+                runtime: rt,
+            },
+            rot,
+        ),
+    })
+}
+
+pub trait DocumentOps {
```

Comment:
Duplicating these traits and all the docs to remove the `async` from the functions seems sub-optimal. Assuming there aren't better options, we could use a procedural macro to generate these traits. Maybe someone else can think of a better option?

Reply:
I briefly looked into an alternative but couldn't find anything. I'm definitely open to alternatives, but I'd need help getting started.

Reply:
In general, when you run out of options to reduce duplication in Rust, procedural macros are the way to go. We leverage the "function-like" variant fairly heavily in both gridiron and recrypt. Here you might be able to use the "attribute" variant here. 

https://doc.rust-lang.org/reference/procedural-macros.html

Reply:
As we decided, I removed the traits, put the functions into `impl BlockingIronOxide` and added hyperlinks to the async docs for all the functions.